File: CIS194_intro_haskell/2013_spring/hw1/card.erl
-module(card).

-export([validate/1]).

%% validate(4012888888881881) is true

-spec validate(integer()) -> boolean().
validate(CardNum) ->
    Digits = to_digit_list(CardNum, []),
    Digits2 = double_odd_positions(Digits, odd, []),
    Sum = sum_digits(Digits2),
    Sum rem 10 == 0.


-spec to_digit_list(integer(), [integer()]) -> [integer()].
to_digit_list(Num, Acc) when Num < 10 -> [Num | Acc];
to_digit_list(Num, Acc) ->
    to_digit_list(Num div 10, [(Num rem 10) | Acc]).


-spec double_odd_positions([integer()], atom(), [integer()]) -> [integer()].
double_odd_positions([], _, Acc) -> lists:reverse(Acc);
double_odd_positions([H | T], odd, Acc) ->
    double_odd_positions(T, even, [(H * 2) | Acc]);
double_odd_positions([H | T], even, Acc) ->
    double_odd_positions(T, odd, [H | Acc]).


-spec sum_digits([integer()]) -> integer().
sum_digits(Digits) ->
    lists:foldl(
      fun
          (N, A) when N > 10 -> sum_digits(to_digit_list(N, [])) + A;
          (N, A) -> N + A
      end,
      0, Digits).

File: CIS194_intro_haskell/2013_spring/hw1/card.hs
{-# OPTIONS_GHC -Wall #-}

validate :: Int -> Bool
validate cardNum =
    let code = (sumDigits . dobleOddPositions . num2digits) cardNum in
    code `rem` 10 == 0


num2digits :: Int -> [Int]
num2digits num =
    num2digits' num []


num2digits' :: Int -> [Int] -> [Int]
num2digits' num acc =
    if num < 10 then num : acc
    else num2digits' d (r : acc)
        where d = num `div` 10
              r = num `rem` 10


dobleOddPositions :: [Int] -> [Int]
dobleOddPositions =
    reverse . snd . foldl f a
        where a = (True, [])
              f = \acc val ->
                  case acc of
                    (True, lst) -> (False, (val * 2) : lst)
                    (False, lst) -> (True, val : lst)


sumDigits :: [Int] -> Int
sumDigits =
    foldl f 0
        where f = \acc val ->
                  if val > 10 then (sumDigits . num2digits) val + acc
                  else val + acc


main :: IO ()
main =
    let cardNum = 4012888888881881 in
    let res = validate cardNum in
    putStrLn $ (show cardNum) ++ " " ++ (show res)
